order-service/src/main/kotlin/cz/kss/proj/orderservice/repository/BaseRepository.kt

```diff
@@ -1,19 +1,40 @@
 package cz.kss.proj.orderservice.repository
 
+import cz.kss.proj.orderservice.model.BaseEntity
+import io.r2dbc.spi.Statement
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.mapNotNull
-import kotlinx.coroutines.reactor.awaitSingle
+import kotlinx.coroutines.reactor.awaitSingleOrNull
 import org.springframework.r2dbc.core.DatabaseClient.GenericExecuteSpec
+import org.springframework.r2dbc.core.awaitRowsUpdated
 
-interface BaseRepository<T, I> {
+interface BaseRepository<T : BaseEntity<I>, I> {
     fun findAll(): Flow<T>
     suspend fun findById(id: I): T?
     suspend fun save(entity: T): T
     suspend fun delete(entity: T): Boolean
     suspend fun deleteAll(): Boolean
 
+    @Suppress("UNCHECKED_CAST")
+    suspend fun insert(entity: T, f: () -> GenericExecuteSpec): T {
+        f().filter { statement: Statement, _ ->
+            statement.returnGeneratedValues("id").execute()
+        }.fetch().first().doOnNext { row ->
+            entity.id = row["id"] as I
+        }.awaitSingleOrNull()
+        return entity
```

Comment:
I am thinking about this approach, If it is good or bad to reuse current object and only add id ?

Reply:
Given, this is an insert operation that literally takes the object "as is" and saves it to the db - I'd argue it's fine. Also, the R2DBC statement doesn't seem to hold the entire object, only the newly created ID. 
If we really want a fresh object, we would have to follow INSERT / UPDATE operations with an immediate SELECT. I'd rather try to avoid this, unless absolutely necessary.